### PR TITLE
RUST-255 Add telemetry on sonar-rust to collect dependencies

### PR DESCRIPTION
**Pull request by thomas-serre-sonarsource**
----
## Summary by Gitar

- **Telemetry enhancement:**
  - Added dependency collection in `Telemetry.java` to track `rust.dependencies` and `rust.dependencies.count`.
  - Implemented `parseDependencies` to extract crates from `Cargo.toml` including standard, dev, build, and target-specific dependencies.
  - Added support for `workspace.dependencies` and dotted key syntax (e.g., `serde.version = "1.0"`).
  - Added truncation logic for the dependency list with random sampling to stay within telemetry limits.
- **Sensor integration:**
  - Updated `RustSensor.java` to invoke dependency telemetry reporting during analysis.
- **Test suite:**
  - Added `TelemetryTest.java` covering edge cases like deduplication, IO errors, truncation, and various `Cargo.toml` syntax formats.
  - Added a test case in `RustSensorTest.java` to verify telemetry integration.

<sub>This will update automatically on new commits.</sub>

### CODE REVIEW

**Comment by thomas-serre-sonarsource** [comment]
This PR is not following the recommendations about having a ProjectSensor rather than a Sensor to report telemetry, but I think this should be done in another dedicated PR. For the time, it's fine if we lose data about dependencies used by our customers. We want to have a glimpse at the most used dependencies, not an exact view.

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/clippy/ClippySensor.java [hunk 1]
```diff
@@ -85,6 +86,7 @@ public void execute(SensorContext context) {
 
     var baseDir = context.fileSystem().baseDir().toPath();
 
+    Telemetry.reportDependencies(context, manifests.stream().map(File::toPath).toList());
```
Comment (marc-jasper-sonarsource) [inline sonar-rust-plugin/src/main/java/org/sonarsource/rust/clippy/ClippySensor.java]: This sits inside `execute` of the `ClippySensor`, meaning that it is never run if someone sets `sonar.rust.clippy.enabled=false`. Is there a way to also collect the dependency data when users decide to disable this?

Comment (marc-jasper-sonarsource) [reply sonar-rust-plugin/src/main/java/org/sonarsource/rust/clippy/ClippySensor.java]: Doesn't this also currently mean that we would not gather this telemetry when importing external Clippy reports?

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java [hunk 2]
```diff
@@ -90,6 +130,132 @@ private static String findRustEdition(Path cargoManifest) throws IOException {
...
+  }
+
+  private static boolean isDependencyTable(String section) {
+    if (DEP_SECTIONS.contains(section)) {
+      return true;
+    }
+    // Target-specific dependencies, e.g. `target.'cfg(unix)'.dependencies`.
+    return section.startsWith("target.")
+      && (section.endsWith(".dependencies") || section.endsWith(".dev-dependencies") || section.endsWith(".build-dependencies"));
+  }
```
Comment (marc-jasper-sonarsource) [inline sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java:236]: This doesn't recognize [workspace.dependencies] or [workspace.dependencies.<crate>], so virtual workspace roots (very common Rust layout) will report incomplete dependencies. Can we add workspace.dependencies here too?

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java [hunk 3]
```diff
@@ -90,6 +130,132 @@ private static String findRustEdition(Path cargoManifest) throws IOException {
...
+      return value;
+    }
+    int limit = MAX_VALUE_LENGTH - TRUNCATION_MARKER.length();
+    int cut = value.lastIndexOf(',', limit);
+    if (cut < 0) {
+      // A single token longer than the limit; hard cut.
+      cut = limit;
+    }
+    return value.substring(0, cut) + TRUNCATION_MARKER;
+  }
```
Comment (marc-jasper-sonarsource) [inline sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java]: I understand that we are facing a limitation here, but this truncation means cutting off based on lexicographical ordering. This can over-represent dependencies which come earlier in the alphabet. Maybe use random sampling if the max. length would be exceeded?

File: sonar-rust-plugin/src/test/java/org/sonarsource/rust/plugin/TelemetryTest.java
Comment (marc-jasper-sonarsource) [inline sonar-rust-plugin/src/test/java/org/sonarsource/rust/plugin/TelemetryTest.java:1]: Could we add an assertion here (or a new test) verifying `sensorContext.addTelemetryProperty("rust.dependencies", ...)` gets called when the sensor runs?

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java [hunk 4]
```diff
@@ -90,6 +130,132 @@ private static String findRustEdition(Path cargoManifest) throws IOException {
...
+    if (matcher.matches()) {
+      String crate = unquote(matcher.group(1).trim());
+      // A sub-table may declare no version; record the crate name up front.
+      deps.putIfAbsent(crate, "");
+      return new Section(SectionKind.DEP_SUBTABLE, crate);
+    }
+    return Section.OTHER;
+  }
+
+  private static void parseEntry(String trimmed, Section section, LinkedHashMap<String, String> deps) {
```
Comment (marc-jasper-sonarsource) [inline sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java:203]: Nit: This doesn't handle dotted-key syntax, e.g. [dependencies] + serde.version = "1.0"

**Review by marc-jasper-sonarsource — CHANGES_REQUESTED** [review]
Nice feature, looking forward to the insights that come from such data. Please see inline comments. I could see improvements regarding how this telemetry is triggered (not disabled when disabling Clippy analysis), regarding the truncation strategy, and how Cargo workspaces are handled.

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/clippy/ClippySensor.java (hunk 1 above)
Comment (thomas-serre-sonarsource) [reply sonar-rust-plugin/src/main/java/org/sonarsource/rust/clippy/ClippySensor.java]: Thanks for spotting that! Moved `reportDependencies` into RustSensor, which runs on every Rust analysis independent of clippy.enabled and of the external report import.

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java (hunk 2 above)
Comment (thomas-serre-sonarsource) [reply sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java:236]: Didn't know about this syntax. Thanks !  I ve updated the code to take care of it.

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java (hunk 3 above)
Comment (thomas-serre-sonarsource) [reply sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java]: random sampling used :+1:

File: sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java (hunk 4 above)
Comment (thomas-serre-sonarsource) [reply sonar-rust-plugin/src/main/java/org/sonarsource/rust/plugin/Telemetry.java:203]: solved !

**Review by marc-jasper-sonarsource — APPROVED** [review]
LGTM! Thanks for addressing all comments.
